src/blockdata/script.rs

```diff
@@ -227,18 +243,66 @@ pub fn read_scriptbool(v: &[u8]) -> bool {
 }
 
 /// Read a script-encoded unsigned integer
+///
+/// ## Errors
+///
+/// This function returns an error in these cases:
+///
+/// * `data` is shorter than `size` - `EarlyEndOfScript`
+/// * `size` is greater than `u16::max_value / 8` (8191) - `NumericOverflow`
+/// * The number being read overflows `usize` - `NumericOverflow`
+///
+/// Note that this does **not** return an error for `size` between `core::size_of::<usize>()`
+/// and `u16::max_value / 8` if there's no overflow.
 pub fn read_uint(data: &[u8], size: usize) -> Result<usize, Error> {
+    read_uint_iter(&mut data.iter(), size).map_err(Into::into)
+}
+
+// We internally use implementation based on iterator so that it automatically advances as needed
+// Errors are same as above, just different type.
+fn read_uint_iter(data: &mut ::core::slice::Iter<'_, u8>, size: usize) -> Result<usize, UintError> {
     if data.len() < size {
-        Err(Error::EarlyEndOfScript)
+        Err(UintError::EarlyEndOfScript)
+    } else if size > usize::from(u16::max_value() / 8) {
```

Comment:
maybe even be more strict with `size > 8` since we haven't arch greater than 64 bits?

Reply:
I agree. This line is clever but we probably don't get any extra generality vs just using 8 as a bound.

Reply:
Was thinking about it but felt wrong to return `NumericOverflow` for what is actually too large input. Was also thinking of

```rust
enum UintSize {
    OneByte = 1,
    TwoBytes = 2,
    FourBytes = 4,
    EightBytes = 8,
}
```

This enforces sane values but is API-breaking and a bit cumbersome. Maybe worth it considering most uses will use literal anyway?

(Also, maybe "width" is better than "size"?)